File: backend/accounts/permissions.py
from rest_framework import permissions


class IsAdmin(permissions.BasePermission):
    """Allow only ADMIN role"""

    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.role == "ADMIN"


class IsAuthor(permissions.BasePermission):
    """Allow only AUTHOR role"""

    def has_permission(self, request, view):
        return request.user.is_authenticated and request.user.role == "AUTHOR"


class IsAdminOrAuthor(permissions.BasePermission):
    """Allow ADMIN or AUTHOR"""

    def has_permission(self, request, view):
        return (
            request.user.is_authenticated
            and request.user.role in ["ADMIN", "AUTHOR"]
        )


class IsAuthorOrAdminForObject(permissions.BasePermission):
    """Object-level permission:
       - Admin can edit/delete any blog
       - Author can edit/delete only their own blog
    """

    def has_object_permission(self, request, view, obj):
        if not request.user.is_authenticated:
            return False
        if request.user.role == "ADMIN":
            return True
        return obj.author == request.user

class IsOwnerOrAdminForObject(permissions.BasePermission):
    """
    Object-level permission:
      - Admin can edit/delete any object
      - Owner (user) can edit/delete their own object
    """
    def has_object_permission(self, request, view, obj):
        if not request.user.is_authenticated:
            return False
        if request.user.role == "ADMIN":
            return True
        # Works for Comment, Like, or any model with 'user' field
        return getattr(obj, "user", None) == request.user

File: frontend/src/utils/permissions.js
// utils/permissions.js

// Role checks
export const isAdmin = (user) => user?.role === "ADMIN";
export const isAuthor = (user) => user?.role === "AUTHOR";
export const isUser = (user) => user?.role === "USER";
export const isAdminOrAuthor = (user) => ["ADMIN", "AUTHOR"].includes(user?.role);



// Object-level permissions
export const canEditOrDeleteBlog = (user, blog) => {
    if (!user) return false;
    if (isAdmin(user)) return true;
    if (isAuthor(user)) return blog.author?.id === user.id;
    return false;
};

export const canTogglePublish = (user, blog) => {
    return canEditOrDeleteBlog(user, blog); // same as edit/delete
};

export const canEditOrDeleteObject = (user, obj) => {
    if (!user) return false;
    if (isAdmin(user)) return true;
    return obj.userId === user.id; // for Comment, Like, etc.
};
